src/Mailgun/Mailgun.php

```diff
@@ -36,17 +43,79 @@ class Mailgun
     protected $apiKey;
 
     /**
-     * @param string|null $apiKey
-     * @param HttpClient  $httpClient
-     * @param string      $apiEndpoint
+     * @var HttpMethodsClient
+     */
+    private $httpClient;
+
+    /**
+     * @var ResponseSerializer
+     */
+    private $serializer;
+
+    /**
+     * @param string|null             $apiKey
+     * @param HttpClient|null         $httpClient
+     * @param string                  $apiEndpoint
+     * @param ResponseSerializer|null $serializer
      */
     public function __construct(
         $apiKey = null,
         HttpClient $httpClient = null,
-        $apiEndpoint = 'api.mailgun.net'
+        $apiEndpoint = 'api.mailgun.net',
+        ResponseSerializer $serializer = null
     ) {
         $this->apiKey = $apiKey;
+        $this->serializer = $serializer ?: new ObjectSerializer();
         $this->restClient = new RestClient($apiKey, $apiEndpoint, $httpClient);
+
+        $plugins = [
+            new Plugin\AddHostPlugin(UriFactoryDiscovery::find()->createUri('https://api.mailgun.net')),
+            new Plugin\HeaderDefaultsPlugin([
+                'User-Agent' => 'mailgun-sdk-php/v2 (https://github.com/mailgun/mailgun-php)',
+                'Authorization' => 'Basic '.base64_encode(sprintf('api:%s', $apiKey)),
+            ]),
+        ];
+
+        $this->httpClient = new HttpMethodsClient(
+            new PluginClient($httpClient ?: HttpClientDiscovery::find(), $plugins),
+            MessageFactoryDiscovery::find()
+        );
+    }
+
+    /**
+     * @param string $name
+     *
+     * @throws \InvalidArgumentException
+     *
+     * @return mixed
+     */
+    public function api($name)
```

Comment:
Github uses this. I guess it is because they want to allow something like: 

```php
$mailgun->getDomainApi();
$mailgun->getDomainsApi();
```

Magic methods will also reduce the number of public methods one have to write. 

Im not sure what to do. I like the current implementation but I understand it is a matter of taste.